src/test/scala/org/neo4j/spark/DataSourceReaderTSE.scala

```diff
@@ -351,6 +348,110 @@ class DataSourceReaderTSE extends SparkConnectorScalaBaseTSE {
+      .option("partitions", "5")
+      .load()
+
+    assertEquals(5, partitionedDf.rdd.getNumPartitions)
+    assertEquals(100, partitionedDf.collect().map(_.getAs[Long]("id")).toSet.size)
+  }
+
+  @Test
+  def testReadRelsCustomPartitions(): Unit = {
+    val fixtureQuery: String =
+      """UNWIND range(1,100) as id
+        |CREATE (p:Person {id: id, name: 'Person ' + id})-[:BOUGHT{quantity: ceil(rand() * 100)}]->(:Product{id: id, name: 'Product ' + id})
+        |RETURN *
+    """.stripMargin
+    SparkConnectorScalaSuiteIT.driver.session()
+      .writeTransaction(
+        new TransactionWork[ResultSummary] {
+          override def execute(tx: Transaction): ResultSummary = tx.run(fixtureQuery).consume()
+        })
+
+    val partitionedDf = ss.read.format(classOf[DataSource].getName)
+      .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
+      .option("relationship", "BOUGHT")
+      .option("relationship.source.labels", ":Person")
+      .option("relationship.target.labels", ":Product")
+      .option("partitions", "5")
+      .load()
+
+    assertEquals(5, partitionedDf.rdd.getNumPartitions)
+    partitionedDf.show()
+    assertEquals(100, partitionedDf.collect().map(_.getAs[Long]("<rel.id>")).toSet.size)
+  }
+
+  @Test
+  def testReadQueryCustomPartitions(): Unit = {
+    val fixtureProduct1Query: String =
+      """CREATE (pr:Product{id: 1, name: 'Product 1'})
+        |WITH pr
+        |UNWIND range(1,100) as id
+        |CREATE (p:Person {id: id, name: 'Person ' + id})-[:BOUGHT{quantity: ceil(rand() * 100)}]->(pr)
+        |RETURN *
+    """.stripMargin
+    SparkConnectorScalaSuiteIT.driver.session()
+      .writeTransaction(
+        new TransactionWork[ResultSummary] {
+          override def execute(tx: Transaction): ResultSummary = tx.run(fixtureProduct1Query).consume()
+        })
+    val fixtureProduct2Query: String =
+      """CREATE (pr:Product{id: 2, name: 'Product 2'})
+        |WITH pr
+        |UNWIND range(1,50) as id
+        |CREATE (p:Person {id: id, name: 'Person ' + id})-[:BOUGHT{quantity: ceil(rand() * 100)}]->(pr)
+        |RETURN *
+    """.stripMargin
+    SparkConnectorScalaSuiteIT.driver.session()
+      .writeTransaction(
+        new TransactionWork[ResultSummary] {
+          override def execute(tx: Transaction): ResultSummary = tx.run(fixtureProduct2Query).consume()
+        })
+
+    val partitionedDf = ss.read.format(classOf[DataSource].getName)
+      .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
+      .option("query",
+        """
+          |MATCH (p:Person)-[r:BOUGHT]->(pr:Product)
+          |RETURN p.name AS person, pr.name AS product, r.quantity AS quantity""".stripMargin)
+      .option("partitions", "5")
+      .load()
+
+    assertEquals(5, partitionedDf.rdd.getNumPartitions)
+    assertEquals(150, partitionedDf.collect().map(_.getAs[Long]("person")).toSet.size)
+
+    val partitionedQueryCountDf = ss.read.format(classOf[DataSource].getName)
```

Comment:
Add tests:

- What happens when query.count is invalid, or fails to return a "count" field, or returns a "count" field that has nonsense?

Reply:
We provided a check if was a READ_ONLY query, I added a check about the returned identifier, that must be only `count`.
What do you mean nonsense? About type? That will be a runtime error because in my knowledge

Reply:
github seems trucated my comment 😥
... in my knowledge there is no way to get the return type from the explain